src/relay/pass/type_solver.cc

```diff
@@ -0,0 +1,166 @@
+/*!
+ *  Copyright (c) 2018 by Contributors
+ * \file type_solver.cc
+ * \brief Type solver implementations.
+ */
+#include <string>
+#include "type_solver.h"
+
+namespace tvm {
+namespace relay {
+
+class TypeSolver::Reporter : public TypeReporterNode {
+ public:
+  explicit Reporter(TypeSolver* solver)
+      : solver_(solver) {}
+
+  void Assign(const Type& dst, const Type& src) final {
+    solver_->Unify(dst, src);
+  }
+
+  void AssertEQ(const ShapeExpr& lhs, const ShapeExpr& rhs) final {
+    // TODO(tqchen)
+  }
+
+ private:
+  TypeSolver* solver_;
+};
+
+// constructor
+TypeSolver::TypeSolver()
+    : reporter_(make_node<Reporter>(this)) {
+}
+
+// destructor
+TypeSolver::~TypeSolver() {
+  // call destructor of all non-POD arena object
+  for (TypeNode* ptr : type_nodes_) {
+    ptr->~TypeNode();
+  }
+  for (RelationNode* ptr : rel_nodes_) {
+    ptr->~RelationNode();
+  }
+}
+
+// Add equality constraint
+Type TypeSolver::Unify(const Type& dst, const Type& src) {
```

Comment:
This doesn't actually implement first order unification on types but just the equivalent of the `Union` operation on the union find. In order for type inference to work we must implement all the functionality from the previous unifier. 

For example I don't see how this handles this case `(?A, ?B) -> ?C = (Tensor<(10, 10), ?bt>, ?W) -> ?Z`, in order to have inference work we need to be able to handle these cases, this is why we had the previous type unifier class.

Reply:
Could the case you gave as an example be handled by going down the AST of the function type and generating equality constraints for each branch? Should that be an updated version of the current unifier?

Reply:
Oh I see the point, owing to your comment on the main thread: The example you gave requires nested incomplete types and is impossible to give without nested incomplete types